### PR TITLE
Implemented a TTL Policy / Ingress restriction test.

### PR DESCRIPTION
**Pull request by alxmrs**
- Refactored the particles to make it easier to trigger reads and writes.
- Added a TTL policy tests that exercises data expiring, egress exfiltration, and data being stored in RAM Disk.

### CODE REVIEW

File: javatests/arcs/android/integration/policy/Particles.kt [hunk 1]
```diff
@@ -15,25 +15,29 @@ import kotlinx.coroutines.Job
...
   val handleRegistered = Job()
-  val outputForTest = mutableSetOf<Thing>()
 
   override fun onReady() {
-    outputForTest.addAll(handles.output.fetchAll())
+    triggerRead()
     handleRegistered.complete()
   }
+
+  fun triggerRead(): Set<Thing> = handles.output.fetchAll()
```
Comment (bgogul) [inline javatests/arcs/android/integration/policy/Particles.kt:42]:
```suggestion
  fun fetchThings(): Set<Thing> = handles.output.fetchAll()
```

File: javatests/arcs/android/integration/policy/Particles.kt [hunk 2]
```diff
@@ -15,25 +15,29 @@ import kotlinx.coroutines.Job
...
 class EgressAB : AbstractEgressAB() {
   val handleRegistered = Job()
-  val outputForTest = mutableSetOf<Thing>()
 
   override fun onReady() {
-    outputForTest.addAll(handles.output.fetchAll())
+    triggerRead()
     handleRegistered.complete()
   }
+
```
Comment (bgogul) [inline javatests/arcs/android/integration/policy/Particles.kt:41]: Not for this PR, the name of the read handle in `EgressAB` should be something other than `output`.

File: javatests/arcs/android/integration/policy/PolicyTest.kt [hunk 3]
```diff
@@ -80,27 +82,90 @@ class PolicyTest {
...
 
-  private suspend fun assertOnlySingletonAbIsPersisted(startingKey: StorageKey) {
+  /**
+   * Scenario: A policy-compliant recipe with @ttl handles egresses data and stores
+   * ingress-restricted values to RAM that are deleted after the TTL expires and at the runtime’s
+   * end.
+   */
+  @Test
+  fun ttlHandlesWithEgress_StoresIngressRestrictedValues_deletedAtTtlTimeAndRuntimeEnd() =
+    runBlocking {
```
Comment (bgogul) [inline javatests/arcs/android/integration/policy/PolicyTest.kt:92]: Is `runBlocking` OK in tests?

File: javatests/arcs/android/integration/policy/PolicyTest.kt [hunk 4]
```diff
@@ -80,27 +82,90 @@ class PolicyTest {
...
+      )
+
+      // When the Arc is run...
+      val arc = env.startArc(TtlEgressesPlan)
+      env.waitForIdle(arc)
+
+      val ingest = env.getParticle<IngressThing>(arc)
+      val egressAB = env.getParticle<EgressAB>(arc)
+
+      withTimeout(30000) {
```
Comment (bgogul) [inline javatests/arcs/android/integration/policy/PolicyTest.kt:105]: I might have asked this earlier, but I can't remember what the answer was. Is using a timeout OK? Isn't there a better way to wait for completion?

File: javatests/arcs/android/integration/policy/PolicyTest.kt [hunk 5]
```diff
@@ -80,27 +82,90 @@ class PolicyTest {
...
+      val startingKey = (egressAB.handles.output.getProxy().storageKey as ReferenceModeStorageKey)
+        .storageKey
+
+      // And only Thing {a, b} is written to storage (RAM)
+      assertOnlySingletonAbIsPersisted(startingKey)
+      env
+
+      // And the egress data with fields Thing {a, b} will be exfiltrated (filtered with no output
+      //  read) after 2 hours have passed.
+      ingest.triggerWrite()
```
Comment (bgogul) [inline javatests/arcs/android/integration/policy/PolicyTest.kt:122]: Why do you need to write again?

File: javatests/arcs/android/integration/policy/PolicyTest.kt [hunk 6]
```diff
@@ -80,27 +82,90 @@ class PolicyTest {
...
+
+      // And Thing {a, b} data will be deleted at Arcs' runtime end
+      assertThat(env.getDatabaseEntities(startingKey, AbstractIngressThing.Thing.SCHEMA)).isEmpty()
+    }
+
+  private suspend fun assertOnlySingletonAbIsPersisted(startingKey: StorageKey) =
+    assertIsPresent(startingKey, setOf("a", "b"))
+
+  /** Assert that all entities only contains values for the specified fields. */
+  private suspend fun assertIsPresent(
```
Comment (bgogul) [inline javatests/arcs/android/integration/policy/PolicyTest.kt:153]:
```suggestion
  private suspend fun assertStorageContains(
```

File: javatests/arcs/android/integration/policy/PolicyTest.kt [hunk 7]
```diff
@@ -80,27 +82,90 @@ class PolicyTest {
...
+
+      // When new data is ingressed...
+      ingest.triggerWrite()
+      withTimeout(30000) {
+        ingest.storeFinished.join()
+      }
+      env.stopArc(arc)
+
+      // Then Thing {a, b} data persists after the arc is finished
+      assertOnlySingletonAbIsPersisted(startingKey)
```
Comment (bgogul) [inline javatests/arcs/android/integration/policy/PolicyTest.kt:141]: Call `assertInPresent`directly.

File: javatests/arcs/android/integration/policy/PolicyTest.kt [hunk 8]
```diff
@@ -80,27 +82,90 @@ class PolicyTest {
...
+
+      env.stopRuntime()
+
+      // And Thing {a, b} data will be deleted at Arcs' runtime end
+      assertThat(env.getDatabaseEntities(startingKey, AbstractIngressThing.Thing.SCHEMA)).isEmpty()
+    }
+
+  private suspend fun assertOnlySingletonAbIsPersisted(startingKey: StorageKey) =
```
Comment (bgogul) [inline javatests/arcs/android/integration/policy/PolicyTest.kt:149]: Let us remove this function. It is making the tests less readable.

File: javatests/arcs/android/integration/policy/scenarios.arcs [hunk 9]
```diff
@@ -38,3 +43,31 @@ recipe PersistsEgresses
...
+    a,
+    b
+  }
+}
+
+// Given a recipe is compliant with a policy
+@policy('SimplePersistencePolicy')
+recipe TtlEgresses
+  // And the recipe's store is marked @ttl('2h')
+  handle1: create @ttl('2h')
```
Comment (bgogul) [inline javatests/arcs/android/integration/policy/scenarios.arcs:64]: Create another handle with `1h` ttl? That way you can run a test to like this: 
  - Write data to two handles.
  - trigger clean up after an hour: data in 1h ttl is gone, but data with 2h ttl is still present
  - trigger clean up after another hour: all data is gone.

File: javatests/arcs/android/integration/policy/PolicyTest.kt (hunk 3 above)
Comment (alxmrs) [reply javatests/arcs/android/integration/policy/PolicyTest.kt:92]: Yes, this is the standard from what I can tell.

File: javatests/arcs/android/integration/policy/PolicyTest.kt (hunk 4 above)
Comment (alxmrs) [reply javatests/arcs/android/integration/policy/PolicyTest.kt:105]: Yeah, I think we opted to keep the timeout in the last discussion. This is just a safeguard, the ingestion should happen right away (unless something is going seriously wrong).

File: javatests/arcs/android/integration/policy/PolicyTest.kt (hunk 5 above)
Comment (alxmrs) [reply javatests/arcs/android/integration/policy/PolicyTest.kt:122]: Looks like this is redundant. Thanks!

File: javatests/arcs/android/integration/policy/scenarios.arcs (hunk 9 above)
Comment (alxmrs) [reply javatests/arcs/android/integration/policy/scenarios.arcs:64]: I like it.

**Comment by alxmrs** [comment]
I think my PolicyTest is flaky. It also may have something to do with the IntegrationEnvironment related flakes mentioned in chat. I'll investigate further.

**Review by bgogul — APPROVED** [review]

**Pull request closed without merging**